### ISSUE TITLE
Not able to reduce loss when training on face and person dataset. Need help. 

### ISSUE THREAD

**Issue by govindamagrawal** [AlexeyAB/darknet#4895]
I am training on the dataset to detect face and person at the same time. The 'person' dataset is from the VOC and COCO training dataset. The 'face' dataset is WIDERFACE dataset. So, there are 2 classes: class 0 as Person and class 1 as Face. The dataset is huge. I am using Tiny-model: 3 yolo layers: https://raw.githubusercontent.com/AlexeyAB/darknet/master/cfg/yolov3-tiny_3l.cfg as recommended for training for both small and large objects. (small faces and large persons). But, when I am training, the average loss is not converging to around 0. It is converging to around 2.5. Also, when  I am doing the detection, it is just showing the face bounding box and not for persons. 
Also, the loss is like:
v3 (mse loss, Normalizer: (iou: 0.750000, **cls: 1.000000**) Region 23 Avg (IOU:0.686603, GIOU: 0.665733), Class: 0.975082, Obj: 0.151904, No Obj: 0.000383, .5R: 0.923077, .75R: 0.230769, count: 13, loss: 2.997764, class_loss = 2.243633, iou_loss = 0.754131

The cls is always equal to 1.000000. Does that mean that it is only training on class 1. When detection in the image, it is only detecting faces( class 1) and no person. 
I have trained for more than 10000 iterations but still could not understand why it the loss is not decreasing and also not detecting the other class.

**Comment by PaserSRL** [comment]
I'm training yolov3-tiny-prn with coco2014 dataset with only 6 classes and after 113000 iterations my avg loss is near 1.7-2.0 
![chart](https://user-images.githubusercontent.com/34647851/74946638-e6102000-53f9-11ea-8307-042718dc7e9f.png)

You need to consider also mAP to understand if training is working properly, what is your value?
Default yolo3-tiny-prn has 33.1% and I've already reached 35.7%

**Comment by shenjifeng** [comment]
I have encoutered the same problem. I was wondering whether there is something wrong with the cfg file.

**Comment by KarthickMani87** [comment]
Hi @govindamagrawal , @shenjifeng ,

I am facing the same issue. Did anyone of you find the answer?

Warm Regards,
Karthick

**Comment by stephanecharette** [comment]
Did you review your annotations?  Most likely you have small annotations where once the image is resized to the network dimensions, the objects (faces) become impossibly small.

**Comment by KarthickMani87** [comment]
Hi @stephanecharette ,

Yes, many of the face annotations are very small. When it was reduced to 416x416 network dimensions, the labels are 4x8, 6x8 (to mention a few). Should I consider annotations only greater than 16x16? I tried upsampling by changing layer from 54 to 23 and increasing strides, but that did not help either.

Looking forward to your response.

Warm Regards,
Karthick

**Comment by stephanecharette** [comment]
See https://www.ccoderun.ca/darkhelp/api/Tiling.html and https://www.ccoderun.ca/darkmark/ImageSize.html

**Comment by KarthickMani87** [comment]
Hi @stephanecharette 

Sorry for the naive questions. Can I use darkhelp to train darknet?

The page explains how to build darknet and darkhelp. But how do I train darknet with darkhelp (with tile enabled)?

Warm Regards,
Karthick

**Comment by stephanecharette** [comment]
Select the "tile images" option in DarkMark when you export the Darknet project:  https://www.ccoderun.ca/darkmark/Summary.html#DarkMarkDarknet

**Comment by stephanecharette** [comment]
See this video for example:  https://www.youtube.com/watch?v=8Ms9T9Ue2g8

**Comment by KarthickMani87** [comment]
Hi @stephanecharette ,

Thank you for the link. I will try these options and get back to you.

Warm Regards,
Karthick

**Comment by KarthickMani87** [comment]
Hi @stephanecharette ,

I am getting the following error:

root@ip-172-31-45-115:~# DarkMark add=/root/Mask/
Invalid MIT-MAGIC-COOKIE-1 keyInvalid MIT-MAGIC-COOKIE-1 key2021-12-21 07:15:00 starting DarkMark v1.6.10-2
2021-12-21 07:15:00 configuration file used: /root/.DarkMark/DarkMark.cfg
2021-12-21 07:15:00 parsing parameter: add=/root/Mask/
2021-12-21 07:15:00 aborting due to signal: "Segmentation fault" [signal #11]
2021-12-21 07:15:00 backtrace #0: DarkMark: get_backtrace[abi:cxx11]() +0x53 [0x55806a1fc093]
2021-12-21 07:15:00 backtrace #1: DarkMark: dm::DarkMarkApplication::signal_handler(int) +0x15c [0x55806a1fd21c]
2021-12-21 07:15:00 backtrace #2: /lib/x86_64-linux-gnu/libc.so.6(+0x3f040) [0x7fd2ff5a2040]
2021-12-21 07:15:00 backtrace #3: DarkMark: juce::ComponentBoundsConstrainer::setBoundsForComponent(juce::Component*, juce::Rectangle<int>, bool, bool, bool, bool) +0x296 [0x55806a4d1cd6]
2021-12-21 07:15:00 backtrace #4: DarkMark: juce::ResizableWindow::setResizeLimits(int, int, int, int) +0x85 [0x55806a4f7625]
2021-12-21 07:15:00 backtrace #5: DarkMark: juce::DocumentWindow::DocumentWindow(juce::String const&, juce::Colour, int, bool) +0xaa [0x55806a503dfa]
2021-12-21 07:15:00 backtrace #6: DarkMark: dm::StartupWnd::StartupWnd() +0x64 [0x55806a2ac604]
2021-12-21 07:15:00 backtrace #7: DarkMark: dm::DarkMarkApplication::initialise(juce::String const&) +0xe57 [0x55806a1fe547]
2021-12-21 07:15:00 backtrace #8: DarkMark: juce::JUCEApplicationBase::initialiseApp() +0x4c [0x55806a37ea0c]
2021-12-21 07:15:00 backtrace #9: DarkMark: juce::JUCEApplicationBase::main() +0x2b [0x55806a1f8e5b]
2021-12-21 07:15:00 backtrace #10: /lib/x86_64-linux-gnu/libc.so.6: __libc_start_main +0xe7 [0x7fd2ff584bf7]
2021-12-21 07:15:00 backtrace #11: DarkMark: _start +0x2a [0x55806a1fb00a]
Aborted (core dumped)

Warm Regards,
Karthick

**Comment by KarthickMani87** [comment]
Hi @stephanecharette ,

I am able to create tiled and zoomed images. I will start training and see if these improve detecting them as face as well as a person.

Warm Regards,
Karthick

**Comment by stephanecharette** [comment]
Remember that darknet does not support tiling natively.  You'll have to use DarkHelp to call darknet in a way that support tiling.  For example, see `-T` here:  https://www.ccoderun.ca/darkhelp/api/Parameters.html

Or, if you want to use the API instead of the CLI, see:  https://www.ccoderun.ca/darkhelp/api/classDarkHelp_1_1Config.html#a78bb66f8709fbbd93643da85640fef1a

**Comment by KarthickMani87** [comment]
Hi @stephanecharette ,

I am planning to use the tiled images along with their annotated text files to train darknet.

Or can I train darknet using DarkHelp? Because while I was going through darkhelp, it asks for .weights which we will get only after darknet training? Or am I missing any command to train darknet using darkhelp?

Warm Regards,
Karthick

**Comment by stephanecharette** [comment]
Read the description here:  https://github.com/stephanecharette/DarkHelp#what-is-the-darkhelp-c-api

DarkHelp uses darknet, it does not replace it.  When you used DarkMark to create the darknet project with tiles, it should have also created a script to start the training.  Look at what that script does.

**Comment by KarthickMani87** [comment]
Hi @stephanecharette ,

Thank you for your prompt responses.

Yes, when I used DarkMark to create the darknet project, it created "images_train.sh" which is "darknet detector train".

I have started my darknet training with the script file. I will post my result if it helps detect an object as a 'person' as well as 'face'.


Warm Regards,
Karthick

**Comment by KarthickMani87** [comment]
Hi @stephanecharette ,

I did the training. The results are not what I expected. The model still struggles to detect the object as a person as well as face. 
Here is the sample output: There are 3 classes here: people with masks, without masks, and people.

Here 3 faces are detected only as "no_masks" and one person as "mask" but they are not detected as "person"
![image](https://user-images.githubusercontent.com/94393722/147275604-13e20057-759f-4c2b-96b5-fe0580bbe632.png)

Can this problem be solved? Would you be open to on-demand consultation?

Warm Regards,
Karthick
